File: Problems 901 to 950/Sort Array by Parity.py
# https://leetcode.com/problems/sort-array-by-parity

# Runtime: 71 ms, faster than 99.07% of Python3 online submissions for Sort Array By Parity.
# Excellent runtime, preallocating the list and then 'refilling' it with values seems to be efficient.

# Memory Usage: 14.7 MB, less than 63.70% of Python3 online submissions for Sort Array By Parity.
# Good memory usage, this probably could have been done in-place without pointers, but this is more readable.


class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        
        length = len(nums)
        left, right = 0, length-1
        res = [0]*length
        
        for n in nums:
            if n % 2:
                res[right] = n
                right -= 1
            else:
                res[left] = n
                left += 1
                
                
        return res

File: Problems 901 to 950/Java/Sort Array By Parity.java
class Solution {
    public int[] sortArrayByParity(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            // is LHS odd and RHS even
            if (nums[left] % 2 > nums[right] % 2) {
                int tmp = nums[left];
                nums[left] = nums[right];
                nums[right] = tmp;
            }
            if (nums[left] % 2 == 0) left++;
            if (nums[right] % 2 == 1) right--;
        }

        return nums;
    }
}
